metadataserver/inject/inject_linux.go

```diff
@@ -0,0 +1,309 @@
+// +build linux
+
+package inject
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"path/filepath"
+	"runtime"
+	"time"
+
+	"github.com/Netflix/titus-executor/logger"
+
+	"github.com/vishvananda/netlink"
+	"github.com/vishvananda/netns"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	tocontainerInterfaceName = "tocontainer"
+	toimdsInterfaceName      = "metadataserver"
+)
+
+var (
+	v6ipnet        *net.IPNet
+	v4ipnet        *net.IPNet
+	localhostipnet *net.IPNet
+	localhost      net.IP
+	maxTime        = 14 * time.Second
+)
+
+func init() {
+	var err error
+	_, v6ipnet, err = net.ParseCIDR("fd00:ec2::254/128")
+	if err != nil {
+		panic(err)
+	}
+
+	_, v4ipnet, err = net.ParseCIDR("169.254.169.254/32")
+	if err != nil {
+		panic(err)
+	}
+
+	localhost = net.ParseIP("127.0.0.1")
+
+	_, localhostipnet, err = net.ParseCIDR("127.0.0.1/32")
+	if err != nil {
+		panic(err)
+	}
+}
+
+/*
+1. It saves the current namespace
+2. It opens the container's network namespace at TITUS_PID_1_DIR/ns/net
+3. It calls unshare, and creates an entirely new network namespace (intermediate network namespace).
+4. It creates a veth pair in the intermediate
+5. It adds the IP address 169.254.169.254 to the veth pair
+6. It binds to port 80 in the new netns created
+7. It moves a side of the veth into the container's network namespace
+8. It sets up a route in the container's netns
+9. It changes to the PID ns of the container
+10. It calls the metadata proxy code
+*/
+func Inject(ctx context.Context, pid1dir string, subsequentExe []string) error {
+	ctx, cancel := context.WithTimeout(ctx, maxTime)
+	defer cancel()
+	listener, err := setupNamespaces(ctx, pid1dir)
+	if err != nil {
+		return fmt.Errorf("Could not get listener: %w", err)
+	}
+
+	/*
+		file, err := listener.File()
+		if err != nil {
+			return fmt.Errorf("Could not get listener file: %w", err)
+		}
+
+		err = unix.Dup2(int(file.Fd()), 169)
+		if err != nil {
+			return fmt.Errorf("Could not clone FD %d into 169: %w", file.Fd(), err)
+		}*/
+
```

Comment:
Leftover?

Reply:
Yeah, need to remove the commented out code.